### DISCUSSION TITLE
`qjsc` usage?

### DISCUSSION THREAD

**Discussion by C-Lunn** [quickjs-ng/quickjs#308] [Q&A]
Hi,

How do I use qjsc? I've been an on hours-long rabbit hole trying to improve my Typst compile times -- the Pintora JS extension compiles on every document compile and I feel like I could speed that up by precompiling. 

I downloaded qjsc for linux and I get ENOENT and feel like I'm going insane -- even running with `strace` doesn't help as it doesn't tell me what file I'm missing.
I downloaded qjsc for Windows and the program silently exits with no feedback no matter what arguments I use.

https://bellard.org/quickjs/quickjs.html#qjsc-compiler-1 this documentation tells me that the compiler exists, but not how to use it.

Can someone please guide me? Thanks.

**Comment by saghul** [comment]
Look at the codegen Makefile target as an example.

The first thing it does is compile the repl JS file to a C file that it's then added to the project so the bytecode can be evaluated.

Is this the use case you have?

**Reply by chqrlie** [reply]
`qjsc` basically converts javascript files to a C file containing byte code arrays. This C file must be compiled and linked with the quickjs runtime.  The behavior of `qjsc` was changed in this fork from the original version in https://github.com/bellard/quickjs : the original version would create an executable directly by launching the C compiler with the appropriate arguments.  This functionality was removed from the current version in this fork for portability reasons across various environments with different C compiler chains and options.

**Reply by guest271314** [reply]
@chqrlie

> This functionality was removed from the current version in this fork for portability reasons across various environments with different C compiler chains and options.

So given something like this

```
/* File generated automatically by the QuickJS-ng compiler. */

#include <inttypes.h>

const uint32_t qjsc_permutations_size = 1497;

const uint8_t qjsc_permutations[1497] = {
 0x11, 0x0f, 0x01, 0x1e, 0x70, 0x65, 0x72, 0x6d,
// ...
```

from this source

```
function array_nth_permutation(a, n) {
    var b = a.slice();  // copy of the set
    var len = a.length; // length of the set
    var res;            // return value, undefined
    var i, f;

    // compute f = factorial(len)
// ...
```

how to compile the generated C code to a standalone executable?

**Reply by bnoordhuis** [reply]
Pass `-e` to qjsc, then compile and link the output against libqjs.

**Reply by guest271314** [reply]
@bnoordhuis Having a helluva time trying to do that. I just fetched this repository. I have already built `qjs` and `qjsc` that are one level above the `quickjs-master` repository.

```
gcc -Wall -L ./quickjs-master -I ./quickjs-master -pie permutations.c -o permutations
```

```
gcc -Wall -L ./quickjs-master -I ./quickjs-master -pie permutations.c -o permutations
/usr/bin/ld: /tmp/ccs8K6Co.o: in function `JS_NewCustomContext':
permutations.c:(.text+0x18): undefined reference to `JS_NewContext'
/usr/bin/ld: /tmp/ccs8K6Co.o: in function `main':
permutations.c:(.text+0x48): undefined reference to `JS_NewRuntime'
/usr/bin/ld: permutations.c:(.text+0x5b): undefined reference to `js_std_set_worker_new_context_func'
/usr/bin/ld: permutations.c:(.text+0x67): undefined reference to `js_std_init_handlers'
/usr/bin/ld: permutations.c:(.text+0x77): undefined reference to `js_module_loader'
/usr/bin/ld: permutations.c:(.text+0x84): undefined reference to `JS_SetModuleLoaderFunc'
/usr/bin/ld: permutations.c:(.text+0xa9): undefined reference to `js_std_add_helpers'
/usr/bin/ld: permutations.c:(.text+0xc8): undefined reference to `js_std_eval_binary'
/usr/bin/ld: permutations.c:(.text+0xd4): undefined reference to `js_std_loop'
/usr/bin/ld: permutations.c:(.text+0xe0): undefined reference to `JS_FreeContext'
/usr/bin/ld: permutations.c:(.text+0xec): undefined reference to `js_std_free_handlers'
/usr/bin/ld: permutations.c:(.text+0xf8): undefined reference to `JS_FreeRuntime'
collect2: error: ld returned 1 exit status

```

**Reply by guest271314** [reply]
Here's what I start out with, in local `~/bin`, where `qjsc` is already built to an executable

```
 qjsc -e -o permutations.c permutations.js
```

That yields

```
/* File generated automatically by the QuickJS-ng compiler. */

#include "quickjs-libc.h"

const uint32_t qjsc_permutations_size = 1497;

const uint8_t qjsc_permutations[1497] = {
 0x11, 0x0f, 0x01, 0x1e, 0x70, 0x65, 0x72, 0x6d,
// ...
 0x01, 0x01, 0x00, 0x07, 0x40, 0x02, 0x66, 0x5e,
 0x00,
};

static JSContext *JS_NewCustomContext(JSRuntime *rt)
{
  JSContext *ctx = JS_NewContext(rt);
  if (!ctx)
    return NULL;
  return ctx;
}

int main(int argc, char **argv)
{
  JSRuntime *rt;
  JSContext *ctx;
  rt = JS_NewRuntime();
  js_std_set_worker_new_context_func(JS_NewCustomContext);
  js_std_init_handlers(rt);
  JS_SetModuleLoaderFunc(rt, NULL, js_module_loader, NULL);
  ctx = JS_NewCustomContext(rt);
  js_std_add_helpers(ctx, argc, argv);
  js_std_eval_binary(ctx, qjsc_permutations, qjsc_permutations_size, 0);
  js_std_loop(ctx);
  JS_FreeContext(ctx);
  js_std_free_handlers(rt);
  JS_FreeRuntime(rt);
  return 0;
}
```

I do not have QuickJS install globally, so that 

```
#include "quickjs-libc.h"
```

is probably not finding `quickjs-libc.h` in `~/bin/quickjs-master`.

**Reply by guest271314** [reply]
@bnoordhuis Alright, I got the executable built with 

```
gcc -c -Wall -I./quickjs-master -pie permutations.c -o permutations
```

then 

```
chmod +x permutations
```

Losing the 


```
console.log(array_nth_permutation([0,1,2,3,4], 5));
```

in translation, as nothing is printed to `stdout` (terminal).

**Reply by guest271314** [reply]
Sure would be helpful to have a minimal, complete, working example of a basic compiling with `qjsc` and `clang` or `gcc` that prints something to `stdout` so we know it works.

**Reply by bnoordhuis** [reply]
There's examples/hello.js that gets turned into an executable if you pass `-DBUILD_EXAMPLES=ON` to cmake. Just tried it, works and prints "hello world".

**Reply by guest271314** [reply]
I'm talking about using `qjsc` after we've already built `qjs` and `qjsc`. Outside of the initial build process. The examples in the wild are fleeting to non-extistent. I looked. That's why I'm asking the source.

**Reply by guest271314** [reply]
Sure, we can use Bytecode Alliance's Javy and `wasmtime` to do this, which under the hood uses QuickJS, however, I think OP has a legitimate inquiry here. Do a search for example of using `qjsc` and `clang` or `gcc` in the wild, and you won't find much, if anything, that provides a complete working example.

**Reply by saghul** [reply]
We can add an example to the docs.

Not sure why you are not seeing output, you should...

**Reply by guest271314** [reply]
I thought so, too. I'm building Bellard's QuickJS and testing this write up https://dustin.boston/blog/2404081/ to see if there's any differences.

**Reply by guest271314** [reply]
Hmm. 

```
wget https://bellard.org/quickjs/quickjs-2024-01-13.tar.xz
tar -xJf quickjs-2024-01-13.tar.xz
```

```
cd quickjs-2024-01-13
make
```

```
cp ~/bin/permutations.js ~/bin/quickjs-2024-01-13
qjsc -e -o permutations.c permutations.js
```

```
gcc permutations.c -o permutations
/usr/bin/ld: /tmp/ccieocST.o: in function `JS_NewCustomContext':
permutations.c:(.text+0x18): undefined reference to `JS_NewContext'
/usr/bin/ld: /tmp/ccieocST.o: in function `main':
permutations.c:(.text+0x48): undefined reference to `JS_NewRuntime'
/usr/bin/ld: permutations.c:(.text+0x5b): undefined reference to `js_std_set_worker_new_context_func'
/usr/bin/ld: permutations.c:(.text+0x67): undefined reference to `js_std_init_handlers'
/usr/bin/ld: permutations.c:(.text+0x77): undefined reference to `js_module_loader'
/usr/bin/ld: permutations.c:(.text+0x84): undefined reference to `JS_SetModuleLoaderFunc'
/usr/bin/ld: permutations.c:(.text+0xa9): undefined reference to `js_std_add_helpers'
/usr/bin/ld: permutations.c:(.text+0xc8): undefined reference to `js_std_eval_binary'
/usr/bin/ld: permutations.c:(.text+0xd4): undefined reference to `js_std_loop'
/usr/bin/ld: permutations.c:(.text+0xe0): undefined reference to `JS_FreeContext'
/usr/bin/ld: permutations.c:(.text+0xec): undefined reference to `js_std_free_handlers'
/usr/bin/ld: permutations.c:(.text+0xf8): undefined reference to `JS_FreeRuntime'
collect2: error: ld returned 1 exit status
```

What am I missing here?

**Reply by guest271314** [reply]
[permutations.js.zip](https://github.com/user-attachments/files/17692519/permutations.js.zip)

**Reply by bnoordhuis** [reply]
> What am I missing here?

You must link against libqjs.a. If you use the cmake build, `target_link_libraries(yourexe, qjs)` takes care of that.

**Reply by guest271314** [reply]
@bnoordhuis My understanding was that `qjsc` is a standalone executable that can be used to compile JavaScript to C and that C is then compiled to an executable. E.g., https://dustin.boston/blog/2404081/.

Evidently QUickJS NG changed the behaviour of `qjsc`. 

I'm just asking for a complete working example - outside of using `cmake`. However, if QuickJS NG depends on `cmake` and a Makefile to work, spelling out all of the step somewhere, for example here, wouldn't hurt anything. You can see my failed attempts to get this working above.

**Reply by bnoordhuis** [reply]
Assuming you already built libqjs.a: `qjsc -e -o hello.c hello.js && cc -o hello hello.c libqjs.a` - add -I if quickjs.h lives in another directory.

If you don't have libqjs.a prebuilt: `cc -o hello hello.c quickjs.c quickjs-libc.c libbf.c libunicode.c cutils.c`

**Reply by guest271314** [reply]
Tried both. Neither works, so far.

**Reply by guest271314** [reply]
@bnoordhuis Unfortunately I got locked out of GitHub when I edited my last comment, above. Long story. Suffice to say make sure you have your 2FA credentials and backup plan in order. GitHub will not help you.

I figured it out for Bellard's QuickJS and QuickJS NG with `make -nB` https://gitlab.com/-/snippets/4769826.

Bellard's QuickJS

```
qjsc -e -fno-string-normalize -fno-map -fno-promise -fno-typedarray -fno-typedarray -fno-regexp -fno-json -fno-eval -fno-proxy -fno-date -fno-module-loader -fno-bigint -o hello.c hello.js
cc -g -Wall -MMD -MF hello.o.d -Wno-array-bounds -Wno-format-truncation -fwrapv  -D_GNU_SOURCE -DCONFIG_VERSION=\"2024-02-14\" -DCONFIG_BIGNUM -DHAVE_CLOSEFROM -O2 -c -o hello.o hello.c -I./quickjs
cc -g -o hello hello.o ./quickjs/.obj/quickjs.o ./quickjs/.obj/libregexp.o ./quickjs/.obj/libunicode.o ./quickjs/.obj/cutils.o ./quickjs/.obj/quickjs-libc.o ./quickjs/.obj/libbf.o -lm -ldl -lpthread -I./quickjs
```

QuickJS NG

```
qjsc -e -o hello.c hello.js
cc -g -Wall -MMD -MF hello.o.d -Wno-array-bounds -Wno-format-truncation -fwrapv  -D_GNU_SOURCE -DCONFIG_VERSION=\"2024-02-14\" -DCONFIG_BIGNUM -DHAVE_CLOSEFROM -O2 -c -o hello.o hello.c -I./quickjs
cc -g -o hello hello.o ./quickjs/build/CMakeFiles/qjs.dir/quickjs.c.o ./quickjs/build/CMakeFiles/qjs.dir/libregexp.c.o ./quickjs/build/CMakeFiles/qjs.dir/libunicode.c.o ./quickjs/build/CMakeFiles/qjs.dir/cutils.c.o ./quickjs/build/CMakeFiles/qjs_exe.dir/quickjs-libc.c.o ./quickjs/build/CMakeFiles/qjs.dir/libbf.c.o -lm -ldl -lpthread -I./quickjs
```

**Reply by saghul** [reply]
You can now do this as well: https://quickjs-ng.github.io/quickjs/cli#creating-standalone-executables

**Reply by guest271314** [reply]
I read that earlier today when adding QuickJS-NG here https://gist.github.com/guest271314/bd292fc33e1b30dede0643a283fadc6a. Havn't tried that QuickJS-NG capability out, yet. Cf. Deno, Node.js, Bun (Bun's implementation expects the source files to still be on the machine https://github.com/oven-sh/bun/issues/14676) https://gist.github.com/guest271314/9b1adad3db3deba64e118f844a77bad6. Facebook's Static Hermes https://gitlab.com/-/snippets/4770898, which AFAICT compiles (to C, then native executable, e.g., using `gcc`) without embedding the engine; there's no simple way to read stdin in `hermes` or `shermes`. Deno has the capability to do this 

```
./deno compile -A npm:esvu
./esvu --engines=v8,spidermonkey
rm -rf node_modules
rm -rf esvu
./deno clean
```

**Reply by guest271314** [reply]
Compiling `npm` to a standalone executable https://gist.github.com/guest271314/c9543a19d8ccf72881355b27d0107551. 

I've been using Bytecode Alliance's Javy (QuickJS engine) to compile JavaScript inside of WASM, then back to asm.js format with `wasm2js`, and to C with `wasm2c`. Still trying to figure out how to compile that resulting C to a native executable, due to dependence on `wasi_snapshot_preview1`.

**Reply by guest271314** [reply]
@saghul 

> You can now do this as well: https://quickjs-ng.github.io/quickjs/cli#creating-standalone-executables

This doesn't work

```
 qjs -c test.js -o app --exe qjs
TypeError: cannot read property 'length' of null
    at compileStandalone (<null>:0:1)
```

This does work

```
qjs test.js 9 362879
5 of 23 (0-indexed, factorial 24) => [0,3,2,1]
```

**Reply by saghul** [reply]
Can you share test.js ?

**Reply by saghul** [reply]
Ah wait, I might have messed up error handling in case we can't find the correct executable. Can you try to skip specifying the exe? It will use the current one by default.

**Reply by saghul** [reply]
Yeah thats it. You need to pass the full path to the exe. I'll fix the error message.

**Reply by guest271314** [reply]
`test.js` I have to tweak this a little at the `scriptArgs` part to work for `qjs`, `shermes`, `node`, `deno`, `bun`, et al.

```
// https://stackoverflow.com/a/34238979
function array_nth_permutation(input, n) {
    n = Number(n);
    let a = Array.from({ length: Number(input) }, (_, i) => i);
    let lex = n;
    let b = Array(); // copy of the set a.slice()
    for (let x = 0; x < a.length; x++) {
      b[x] = a[x];
    }
    let len = a.length; // length of the set
    const res = Array(); // return value, undefined
    let i = 1;
    let f = 1;

    // compute f = factorial(len)
    for (; i <= len; i++) {
      f *= i;
    }

    let fac = f;
    // if the permutation number is within range
    if (n >= 0 && n < f) {
      // start with the empty set, loop for len elements
      // let result_len = 0;
      for (; len > 0; len--) {
        // determine the next element:
        // there are f/len subsets for each possible element,
        f /= len;
        // a simple division gives the leading element index
        i = (n - n % f) / f; // Math.floor(n / f);
        // alternately: i = (n - n % f) / f;
        // res[(result_len)++] = b[i];
        // for (let j = i; j < len; j++) {
        //   b[j] = b[j + 1]; // shift elements left
        // }
        res.push(b.splice(i, 1)[0]);
        // reduce n for the remaining subset:
        // compute the remainder of the above division
        n %= f;
        // extract the i-th element from b and push it at the end of res
      }
      return `${lex} of ${fac - 1} (0-indexed, factorial ${fac}) => ${
        JSON.stringify(res)
      }`;
    } else {
      if (n === 0) {
        return `${JSON.stringify(res)}\n`;
      }
      return `${n} >= 0 && ${n} < ${f}: ${n >= 0 && n < f}`;
    }
  }
  
console.log(array_nth_permutation(...scriptArgs.slice(-2)));
```

**Reply by guest271314** [reply]
Yeah, passing `--exe ~/bin/qjs` works. `qjs` is in `PATH` because I have `~/bin` in `PATH`. No way to figure that out with the current error message.
